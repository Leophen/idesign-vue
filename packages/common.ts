import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
