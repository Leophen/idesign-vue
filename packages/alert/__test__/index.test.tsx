import { Alert } from '../index';
import { render } from '../../common'

describe('Alert 组件测试', () => {
  it('type', () => {
    const props = {
      message: 'Alert 内容',
    };
    const wrapper = render(<Alert {...props} type="info" />);
    expect(wrapper.element).toMatchSnapshot();

    const wrapper2 = render(<Alert {...props} type="success" />);
    expect(wrapper2.element).toMatchSnapshot();

    const wrapper3 = render(<Alert {...props} type="warning" />);
    expect(wrapper3.element).toMatchSnapshot();

    const wrapper4 = render(<Alert {...props} type="error" />);
    expect(wrapper4.element).toMatchSnapshot();
  });

  it('closable', () => {
    const wrapper = render(<Alert message="这是一条消息提示" closable />);
    wrapper.find('.i-alert--close-btn').trigger('click');
    expect(wrapper.find('i-alert').exists()).toBe(false);
  })

  it('operation', () => {
    const operation = <span data-test="test">相关操作</span>;
    const wrapper = render(
      <Alert message="这是一条消息提示">
        {{
          default: () => operation,
        }}
      </Alert>
    );
    expect(wrapper.find('[data-test="test"]').text()).toBe('相关操作');
  })

  it('title', () => {
    const wrapper = render(<Alert title="消息提示标题" message="这是一条消息提示" />);
    expect(wrapper.find('.i-alert--title').text()).toBe('消息提示标题');
  })
});



