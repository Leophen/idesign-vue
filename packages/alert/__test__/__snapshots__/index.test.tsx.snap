// Vitest Snapshot v1

exports[`Alert 组件测试 > type 1`] = `
<div
  class="i-alert i-alert--type-info"
  message="Alert 内容"
>
  <i
    class="i-icon icon-TipInfo"
    style="font-size: 16px;"
  />
  <div
    class="i-alert--content"
  >
    <!--v-if-->
    <div
      class="i-alert--description"
    >
      
      
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`Alert 组件测试 > type 2`] = `
<div
  class="i-alert i-alert--type-success"
  message="Alert 内容"
>
  <i
    class="i-icon icon-TipCheckFill"
    style="font-size: 16px;"
  />
  <div
    class="i-alert--content"
  >
    <!--v-if-->
    <div
      class="i-alert--description"
    >
      
      
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`Alert 组件测试 > type 3`] = `
<div
  class="i-alert i-alert--type-warning"
  message="Alert 内容"
>
  <i
    class="i-icon icon-TipWarningFill"
    style="font-size: 16px;"
  />
  <div
    class="i-alert--content"
  >
    <!--v-if-->
    <div
      class="i-alert--description"
    >
      
      
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
</div>
`;

exports[`Alert 组件测试 > type 4`] = `
<div
  class="i-alert i-alert--type-error"
  message="Alert 内容"
>
  <i
    class="i-icon icon-TipCloseFill"
    style="font-size: 16px;"
  />
  <div
    class="i-alert--content"
  >
    <!--v-if-->
    <div
      class="i-alert--description"
    >
      
      
      <!--v-if-->
    </div>
  </div>
  <!--v-if-->
</div>
`;
